app/Services/TestDiffService.php

```diff
@@ -0,0 +1,136 @@
+
+/**
+ * This class is responsible for populating the `testdiff` table.
+ * It also sets `build2test::newstatus` = 1 as appropriate.
+ **/
+
+class TestDiffService
+{
+    public static function computeDifferences(Build $build) : bool
+    {
+        // Populate previous build.
+        $previous_build_id = $build->GetPreviousBuildId();
+        if ($previous_build_id == 0) {
+            return false;
+        }
+        $previous_build = new Build();
+        $previous_build->Id = $previous_build_id;
+        $previous_build->FillFromId($previous_build->Id);
+
+        return self::computeDifferencesForType($build, $previous_build, TestDiff::TEST_TYPE_NOTRUN) &&
+            self::computeDifferencesForType($build, $previous_build, TestDiff::TEST_TYPE_FAILED) &&
+            self::computeDifferencesForType($build, $previous_build, TestDiff::TEST_TYPE_PASSED) &&
+            self::computeDifferencesForType($build, $previous_build, TestDiff::TEST_TYPE_FAILED_TIMESTATUS);
+    }
+
+    private static function computeDifferencesForType(Build $build, Build $previous_build, int $type) : bool
+    {
+        $tests = [];
+        $previous_tests = [];
+
+        switch ($type) {
+            case TestDiff::TEST_TYPE_NOTRUN:
+                $tests = $build->GetNotRunTests();
+                $previous_tests = $previous_build->GetNotRunTests();
+                break;
+            case TestDiff::TEST_TYPE_FAILED:
+                $tests = $build->GetFailedTests();
+                $previous_tests = $previous_build->GetFailedTests();
+                break;
+            case TestDiff::TEST_TYPE_PASSED:
+                $tests = $build->GetPassedTests();
+                $previous_tests = $previous_build->GetPassedTests();
+                break;
+            case TestDiff::TEST_TYPE_NOTRUN:
+                $tests = $build->GetFailedTimeStatusTests();
+                $previous_tests = $previous_build->GetFailedTimeStatusTests();
+                break;
+            default:
+                return false;
+                break;
+        }
+
+        // Number of tests for this build that newly achieved this status.
+        $num_positive = 0;
+
+        // Number of tests for this build that no longer have this status when compared to the previous build.
+        $num_negative = 0;
+
+        // Array of build2test::ids corresponding to tests from this build whose statuses
+        // has changed since the previous build.
+        $newstatus_b2t_ids = [];
+
+        foreach ($tests as $test) {
+            if (array_search($test['name'], array_column($previous_tests, 'name')) === false) {
+                // This test was found for the current build but not the previous build.
+                $num_positive += 1;
+                $newstatus_b2t_ids[] = $test['buildtestid'];
+            }
+        }
+
+        // Set build2test::newstatus=1 for tests whose status has changed since this previous build.
+        \DB::table('build2test')
```

Comment:
It would be better to explicitly `use` `Illuminate\Support\Facades\DB` here since most IDEs and static analysis tools aren't able to see that Laravel is importing them for us.  Apply throughout the remainder of this file.

```suggestion
        DB::table('build2test')
```